File: packages/backend/src/services/note/watch.ts
import type { User } from "@/models/entities/user.js";
import type { Note } from "@/models/entities/note.js";
import { NoteWatchings } from "@/models/index.js";
import { genId } from "backend-rs";
import type { NoteWatching } from "@/models/entities/note-watching.js";

export default async (me: User["id"], note: Note) => {
	// 自分の投稿はwatchできない
	if (me === note.userId) {
		return;
	}

	await NoteWatchings.insert({
		id: genId(),
		createdAt: new Date(),
		noteId: note.id,
		userId: me,
		noteUserId: note.userId,
	} as NoteWatching);
};

File: packages/backend-rs/src/service/note/watch.rs
use crate::database::db_conn;
use crate::model::entity::note_watching;
use crate::util::id::gen_id;
use sea_orm::{prelude::*, ActiveValue};

#[crate::export]
pub async fn watch_note(
    watcher_id: &str,
    note_author_id: &str,
    note_id: &str,
) -> Result<(), DbErr> {
    if watcher_id != note_author_id {
        note_watching::Entity::insert(note_watching::ActiveModel {
            id: ActiveValue::set(gen_id()),
            created_at: ActiveValue::set(chrono::Utc::now().into()),
            user_id: ActiveValue::Set(watcher_id.to_string()),
            note_user_id: ActiveValue::Set(note_author_id.to_string()),
            note_id: ActiveValue::Set(note_id.to_string()),
        })
        .exec(db_conn().await?)
        .await?;
    }

    Ok(())
}

#[crate::export]
pub async fn unwatch_note(watcher_id: &str, note_id: &str) -> Result<(), DbErr> {
    let db = db_conn().await?;

    let entry = note_watching::Entity::find()
        .filter(note_watching::Column::UserId.eq(watcher_id))
        .filter(note_watching::Column::NoteId.eq(note_id))
        .one(db)
        .await?;

    if let Some(entry) = entry {
        entry.delete(db).await?;
    }

    Ok(())
}
